### PR TITLE
prov/efa: Allow apps to reset RNR retry counter

### PR DESCRIPTION
**Pull request by zhngaj**
prov/efa: Allow apps to reset RNR retry counter 

This needs rebase after #7065 is merged.

Update: #7065 merged. This PR has been rebased.

### CODE REVIEW

**Comment by shefty** [comment]
Please separate changes to the main external API into their own PR, so that they don't get lost as part of a efa specific change.

File: include/rdma/fi_endpoint.h [hunk 1]
```diff
@@ -66,6 +66,7 @@ enum {
 	FI_OPT_RECV_BUF_SIZE,
 	FI_OPT_TX_SIZE,
 	FI_OPT_RX_SIZE,
+	FI_OPT_RNR_RETRY,
```
Comment (shefty) [inline include/rdma/fi_endpoint.h:69]: RNR is not an API related definition.  This is a provider internal implementation detail.

Comment (zhngaj) [reply include/rdma/fi_endpoint.h:69]: I will move this to efa provider specific definition.

File: prov/efa/src/rxr/rxr_ep.c [hunk 2]
```diff
@@ -1055,28 +1055,70 @@ static int rxr_ep_getopt(fid_t fid, int level, int optname, void *optval,
...
 
-	*(size_t *)optval = rxr_ep->min_multi_recv_size;
-	*optlen = sizeof(size_t);
+	switch (optname) {
+	case FI_OPT_MIN_MULTI_RECV:
+		*(size_t *)optval = rxr_ep->min_multi_recv_size;
+		*optlen = sizeof(size_t);
+		break;
+	case FI_OPT_RNR_RETRY:
+		*(size_t *)optval = rxr_env.rnr_retry;
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_ep.c:1067]: rxr_env is a global and should not be used to store endpoint state

File: include/rdma/fi_endpoint.h (hunk 1 above)
Comment (zhngaj) [reply include/rdma/fi_endpoint.h:69]: @shefty, if customer expects to reset RNR retry counter via fi_setopt, can we define a range for providers to define their specific options?

Comment (shefty) [reply include/rdma/fi_endpoint.h:69]: Yes, you should be able to use FI_PROV_SPECIFIC to set the high-order bit indicating a provider specific value.

Comment (zhngaj) [reply include/rdma/fi_endpoint.h:69]: @shefty I'm not sure I fully understand your suggestion, so could be easily wrong here. 
Is it something like this? 
```
--- a/include/rdma/fi_endpoint.h
+++ b/include/rdma/fi_endpoint.h
@@ -66,6 +66,7 @@ enum {
        FI_OPT_RECV_BUF_SIZE,
        FI_OPT_TX_SIZE,
        FI_OPT_RX_SIZE,
+       FI_OPT_EFA_RNR_RETRY = (1U | FI_PROV_SPECIFIC),
 };
```
Or we need to create a provider-specific header file, define FI_OPT_EFA_RNR_RETRY in it, and apps need to include this new header file to properly call fi_setopt with the defined optname? Basically follow the similar approach of https://github.com/ofiwg/libfabric/blob/main/prov/usnic/src/fi_ext_usnic.h

Comment (shefty) [reply include/rdma/fi_endpoint.h:69]: Let me respond on the mailing list

File: prov/efa/src/rxr/rxr_ep.c (hunk 2 above)
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_ep.c:1067]: Will have a new field in endpoint struct to store the RNR retry count.

File: include/rdma/fi_endpoint.h (hunk 1 above)
Comment (zhngaj) [reply include/rdma/fi_endpoint.h:69]: Thanks, @shefty. I think we reach to a conclusion on the mailing list. In my case, I will
1. In fi_endpoint.h, define a FI_OPT_PROV_SPECIFIC  (1U << 30)
2. In fi_ext.h, define FI_OPT_EFA_RNR_RETRY (1U | FI_OPT_PROV_SPECIFIC)
Am i understanding correctly? 
If you have a generic example showing as a guideline, pls let me know, I can add it as the comment for other providers to follow.

Comment (shefty) [reply include/rdma/fi_endpoint.h:69]: I would use the existing FI_PROV_SPECIFIC definition, versus also defining FI_OPT_PROV_SPECIFIC.

I was also considering a provider specific code in fi_ext.h that could ensure 2 providers don't select the same values.  E.g.
```
#define FI_PROV_SPECIFIC_EFA (FI_PROV_SPECIFIC | (0xefa << 20))
#define FI_PROV_SPECIFIC_VERBS (FI_PROV_SPECIFIC | 0xd3a << 20)

/* EFA options */
enum {
    FI_OPT_EFA_RNR_RETRY (1U | FI_PROV_SPECIFIC_EFA)
};
```
Not sure if the provider codes should be 8, 12, or 16 bits, or if it really matters much.

Comment (zhngaj) [reply include/rdma/fi_endpoint.h:69]: One thing I think we should consider is that using FI_PROV_SPECIFIC (#define FI_PROV_SPECIFIC (1U << 31)) could exceed the range `int` allows, as optname is type of integer in definition (`fi_setopt(fid_t fid, int level, int optname, const void *optval, size_t optlen)`)

Comment (shefty) [reply include/rdma/fi_endpoint.h:69]: Are we really dealing with any processor that doesn't have a 32-bit integer?

Comment (zhngaj) [reply include/rdma/fi_endpoint.h:69]: No, that's just a minor thought. Let's ignore it. 
For provider codes, 12 or 16 bits looks good to me,  which allows a "meaningful" provider code and also leave large enough space (20 or 16 bits) for provider specific values.

File: prov/efa/src/rxr/rxr_ep.c (hunk 2 above)
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_ep.c]: Fixed in new revision.

File: include/rdma/fi_endpoint.h (hunk 1 above)
Comment (zhngaj) [reply include/rdma/fi_endpoint.h]: changes are in #7065

File: man/fi_efa.7.md [hunk 3]
```diff
@@ -82,6 +82,14 @@ No support for counters for the DGRAM endpoint.
 
 No support for inject.
 
+# PROVIDER SPECIFIC ENDPOINT LEVEL OPTION
+
+*FI_OPT_EFA_RNR_RETRY*
+: Defines the number of RNR retry. The application can use it to reset RNR retry
+  counter via the call to fi_setopt. Note that the reset must be before endpoint
```
Comment (rwespetal) [inline man/fi_efa.7.md:89]: I'd suggest "Note that this option must be set before the endpoint is enabled."

File: prov/efa/src/rxr/rxr_ep.c [hunk 4]
```diff
@@ -1111,31 +1111,76 @@ static ssize_t rxr_ep_cancel(fid_t fid_ep, void *context)
...
+		 * if the call to fi_setopt is before or after EP enabled for
+		 * convience, instead of calling to ibv_query_qp
+		 */
+		if (!efa_ep->qp) {
+			efa_ep->rnr_retry = *(size_t *)optval;
+		} else {
+			FI_WARN(&rxr_prov, FI_LOG_EP_CTRL,
+				"The option FI_OPT_EFA_RNR_RETRY is required \
+				to be set before EP enabled %s\n", __func__);
+			return -FI_EPERM;
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_ep.c:1176]: FI_EINVAL is probably a better return value here

**Review by rwespetal — APPROVED** [review]

File: man/fi_efa.7.md (hunk 3 above)
Comment (zhngaj) [reply man/fi_efa.7.md]: Sure, fixed in new rev.

File: prov/efa/src/rxr/rxr_ep.c (hunk 4 above)
Comment (zhngaj) [reply prov/efa/src/rxr/rxr_ep.c]: Updated in new rev.

**Comment by zhngaj** [comment]
@shefty May I get failure in continuous-integration/jenkins/pr-head if it's a real issue?

**Comment by shefty** [comment]
It looks like it was a hang running MPI on the psm2 provider.

**Comment by zhngaj** [comment]
Then it's not related to the change in this PR. Thanks.

**Review by wzamazon — APPROVED** [review]

**Pull request merged**